### ISSUE TITLE
[GLFW] Wrong compile time if-branch is hit due to the missing definition of `MAC_OS_X_VERSION_MAX_ALLOWED`

### ISSUE THREAD

**Issue by Gnimuc** [JuliaPackaging/Yggdrasil#1733]
Background: https://github.com/JuliaGL/GLFW.jl/issues/207

The following code built with Xcode 12 gives `ok! 101500`. Whereas it prints `error!` if I compile it with BB2 macOS SDK.
```c
#include <stdio.h>

int main(int argc, const char * argv[]) {
#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101100
    printf("ok! %d\n", MAC_OS_X_VERSION_MAX_ALLOWED);
#else
    printf("error!\n");
#endif
    return 0;
}
```

In this specific case, `MAC_OS_X_VERSION_MAX_ALLOWED` is not defined in the BB2 macOS SDK. A simple fix would be to add `-DMAC_OS_X_VERSION_MAX_ALLOWED=101100` to glfw_jll's build script. I'm wondering how to fix this in a more generic way. Maybe add some facilities to let users choose preferred macOS SDK versions like what we currently do for gcc? 

BTW, problems caused by this kind of bugs are not trivial, one probably needs to read the source code to trace the root cause.

**Comment by giordano** [comment]
The macro is defined in `AvailabilityMacros.h`, if you don't load the header file the macro is undefined:
```
sandbox:${WORKSPACE} # cat test.c 
#include <stdio.h>

int main(int argc, const char * argv[]) {
#ifdef MAC_OS_X_VERSION_MAX_ALLOWED
    #if MAC_OS_X_VERSION_MAX_ALLOWED >= 101100
        #warning "MAC_OS_X_VERSION_MAX_ALLOWED recent enough"
    #else
        #error "MAC_OS_X_VERSION_MAX_ALLOWED too old"
    #endif
#else
    #error "Macro MAC_OS_X_VERSION_MAX_ALLOWED undefined"
#endif
    return 0;
}
sandbox:${WORKSPACE} # cc test.c 
test.c:11:6: error: "Macro MAC_OS_X_VERSION_MAX_ALLOWED undefined"
    #error "Macro MAC_OS_X_VERSION_MAX_ALLOWED undefined"
     ^
1 error generated.
sandbox:${WORKSPACE} # cat test.c 
#include <stdio.h>
#include <AvailabilityMacros.h>

int main(int argc, const char * argv[]) {
#ifdef MAC_OS_X_VERSION_MAX_ALLOWED
    #if MAC_OS_X_VERSION_MAX_ALLOWED >= 101100
        #warning "MAC_OS_X_VERSION_MAX_ALLOWED recent enough"
    #else
        #error "MAC_OS_X_VERSION_MAX_ALLOWED too old"
    #endif
#else
    #error "Macro MAC_OS_X_VERSION_MAX_ALLOWED undefined"
#endif
    return 0;
}
sandbox:${WORKSPACE} # cc test.c 
test.c:7:10: warning: "MAC_OS_X_VERSION_MAX_ALLOWED recent enough" [-W#warnings]
        #warning "MAC_OS_X_VERSION_MAX_ALLOWED recent enough"
         ^
1 warning generated.
```
BinaryBuilder now uses SDK 10.12, which should be recent enough for what you want

**Comment by Gnimuc** [comment]
So the proper fix is actually to patch GLFW's source code to `#include <AvailabilityMacros.h>`?

**Comment by giordano** [comment]
I assume it does already, otherwise how would the macro be defined at all?  Maybe just try to rebuild?  We used to use SDK 10.10 before

**Comment by Gnimuc** [comment]
I did a global search for `AvailabilityMacros` over glfw's source repo and it returned nothing. `MAC_OS_X_VERSION_MAX_ALLOWED ` is probably defined in the newer macOS's system headers.

**Comment by giordano** [comment]
Maybe it's included indirectly in header files directly included?  Also

>  Maybe just try to rebuild? We used to use SDK 10.10 before

**Comment by Gnimuc** [comment]
I also did a local build with `-DMAC_OS_X_VERSION_MAX_ALLOWED=101100`. The Vulkan stuff works well with this build. I'm going to give #1734 a test tomorrow. If it doesn't work, I'll submit a bugfix to the upstream.

**Comment by giordano** [comment]
Again, if they aren't including `AvailabilityMacros.h` in any way, be it direct or indirect, this package wouldn't work anywhere.  The simplest explanation is that until a few weeks ago the SDK available was 10.10 (which is older than 10.11), that's why simply rebuilding now should work

**Comment by Gnimuc** [comment]
> The simplest explanation is that until a few weeks ago the SDK available was 10.10 (which is older than 10.11), that's why simply rebuilding now should work

Yes, you're right. I just tested the latest GLFW_jll and it works. The root cause is indeed the old SDK version.
